TNL-4324: Enhance Mako linting

robrap opened the pull request:
## [TNL-4324](https://openedx.atlassian.net/browse/TNL-4324)

Enhance Mako linting:
- Lint JavaScript context for JavaScript violations
- Lint for Mako specific JavaScript rules
- Skip commented lines
- Change unicode to decode.utf8

Other enhancements:
- Count lint violations

### Testing
- [x] Data is properly encoded in HTML templates to avoid XSS
- [x] Unit, integration, acceptance tests as appropriate

### Reviewers
If you've been tagged for review, please check your corresponding box once you've given the :+1:.
- [x] Code review: @efischer19 
- [x] Code review: @nedbat 

### Post-review
- [ ] Squash commits

robrap commented:
run jenkins js

robrap commented:
jenkins run all

nedbat commented on scripts/safe_template_linter.py:
```diff
@@ -1402,10 +1500,64 @@ def _check_mako_expressions(self, mako_template, has_page_default, results):
+        Side effect:
+            Adds JavaScript violations to results.
+        """
+        javascript_start_index = -1
```
It would be more usual in Python to use None to mean "no line yet".

robrap replied:
@nedbat: That makes sense.  These are things a Java developer needs to get used to. :)

nedbat commented on scripts/safe_template_linter.py:
```diff
@@ -1402,10 +1500,64 @@ def _check_mako_expressions(self, mako_template, has_page_default, results):
+        for violation in javascript_results.violations:
+            expression = violation.expression
+            expression['start_index'] += start_offset
+            if 0 < expression['end_index']:
```
Can we use None here also?

efischer19 commented on scripts/safe_template_linter.py:
```diff
@@ -146,6 +146,7 @@ def __init__(self, string):
         """
         self._string = string
         self._line_breaks = self._process_line_breaks(string)
+        self.eof_index = len(string)
```
It could just be me, but a comment here explaining that this is an exclusive index would be helpful.

nedbat commented on scripts/safe_template_linter.py:
```diff
@@ -1521,8 +1674,8 @@ def _check_filters(self, expression, context, has_page_default, results):
         filters = filters_match.group()[1:-1].replace(" ", "").split(",")
-        if (len(filters) == 2) and (filters[0] == 'n') and (filters[1] == 'unicode'):
-            # {x | n, unicode} is valid in any context
+        if (len(filters) == 2) and (filters[0] == 'n') and (filters[1] == 'decode.utf8'):
```
This is easier written as `if filters == ['n', 'decode.utf8']:`

nedbat commented on scripts/safe_template_linter.py:
```diff
@@ -1521,8 +1674,8 @@ def _check_filters(self, expression, context, has_page_default, results):
+            # {x | n, decode.utf8} is valid in any context
             pass
         elif context == 'html':
             if (len(filters) == 1) and (filters[0] == 'h'):
```
`if filters == ['h']:`

nedbat commented on scripts/safe_template_linter.py:
```diff
@@ -1511,7 +1664,7 @@ def _check_filters(self, expression, context, has_page_default, results):
         """
         # finds "| n, h}" when given "${x | n, h}"
-        filters_regex = re.compile(r'\|[a-zA-Z_,\s]*\}')
+        filters_regex = re.compile(r'\|[a-zA-Z0-9_.,\s]*\}')
```
This will match `|}`, not sure if you want it to.

efischer19 replied:
`\w`:

... matches any alphanumeric character and the underscore; this is equivalent to the set [a-zA-Z0-9_].
https://docs.python.org/2/library/re.html

nedbat commented on scripts/safe_template_linter.py:
```diff
@@ -1536,19 +1689,94 @@ def _check_filters(self, expression, context, has_page_default, results):
+        if 0 < expression['end_index']:
+            end_index = lines.index_to_line_end_index(expression['end_index'])
+        else:
+            return None
```
Can the expression really not have an end_index? And if so, it's ok to not return a string?

efischer19 commented:
Overall, looks good - just a few minor things to address. Ned's comments about the use of `None` vs an index of -1 are the big thing that stood out to me.

nedbat commented on scripts/safe_template_linter.py:
```diff
@@ -982,6 +1041,15 @@ def _check_javascript_file_is_safe(self, file_contents, results):
+        """
+        for violation in list(results.violations):
+            if violation.first_line().lstrip().startswith('//'):
+                results.violations.remove(violation)
```
This code is a lot like the other implementation below (just the comment criterion is different).  Maybe refactor?  Also, it won't matter for short lists like this, but iterating and then using remove will be O(n**2).  You could do:
```
results.violations = [v for v in results.violations if not self.is_comment(v)]
```

nedbat commented:
Some Python style points.  :+1:

robrap commented:
@nedbat @efischer19: I added a commit that addresses all issues.  
1. I added some tests for StringLines around that new `eof_index` that @efischer19 asked about, and it turned up a defect in `line_number_to_end_index`.
2. As part of changing all -1 references to None, I created a new "Expression" class.  It is mostly just a dumb container, but it helped me find things.
All tests continue to pass.  I am looking for thumbs again.  Feel free to do a quick pass or just say if you want me to rely on the tests and past thumb.  
Thanks!

robrap commented:
And thank you for the reviews!

nedbat commented:
Latest commit looks like about what I'd expect for those changes :)  :+1:

robrap commented:
Thanks @nedbat. 
FYI: I'm going to merge now without waiting on the tests, and squash and test one-level up on the robrap/linter-move-mako branch: https://github.com/edx/edx-platform/pull/12177.
